`assert failed fs/shim_fs.c:533 dent == dent2` in DEBUG mode only

<!-- Thanks for reporting a bug.

Please provide as much information as possible.
Paste any logs using three backticks (```).

IMPORTANT: please do not post general questions here;
if you have questions, please use mailing list: support@graphene-project.io
-->

## Description of the problem

### Error output

Using the latest master code and build it in DEBUG=1 and SGX=1, there will be an error output like this:

> ...
> enclave initializing:
>     enclave id:   0x000000001ffff000
>     mr_enclave:   6aa49ee92e560d6c50de0f1e3637953edc79e07743fd2972396ef0c4a561300c
> Seal key: ebd1f5bc9e88c88dc786e38e2996d5e4
> trusted: file:pipe
> trusted: file:/home/ya0guang/Code/graphene/Runtime/libsysdb.so
> trusted: file:/home/ya0guang/Code/graphene/Runtime/ld-linux-x86-64.so.2
> trusted: file:/home/ya0guang/Code/graphene/Runtime/libc.so.6
> trusted: file:/lib/x86_64-linux-gnu/libc.so.6
> WARNING: Using insecure argv source. Don't use this configuration in production!
> assert failed fs/shim_fs.c:533 dent == dent2
> DkProcessExit: Returning exit code 1
> make: *** [Makefile:93: run] Error 1
> rm pipe-deps pipe-trusted-libs

But the program works as expected when building without `DEBUG=1`.

### Source code

My sample source code of compiled binary:

```c
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <unistd.h>
#include <string.h>

int main(void)
{
    pid_t pid;
    int i = 0;
    int result = -1;
    int fd[2], nbytes;
    char string[100] = "I'm a pipe!";
    char readbuffer[80];
    int *write_fd = &fd[1];
    int *read_fd = &fd[0];

    printf("Please input data:");
    // scanf("%s", string);
    result = pipe(fd);
    if (-1 == result)
    {
        perror("pipe");
        return -1;
    }

    pid = fork();
    if (-1 == pid)
    {
        perror("fork");
        return -1;
    }
    else if (0 == pid)
    {
        printf("this is child % d\n", getpid());
        close(*read_fd);
        result = write(*write_fd, string, strlen(string));
        return 0;
    }
    else
    {
        printf("this is parent % d\n", getpid());
        close(*write_fd);
        nbytes = read(*read_fd, readbuffer, sizeof(readbuffer) - 1);
        printf("receive %d data：%s\n", nbytes, readbuffer);
    }
    return 0;
}
```

### Normal output

After rebuilding the code by executing `make clean` and `make SGX=1` , it works correctly with output like this:

> ...
> Enclave identifier: 949f4909c927aba8 -> 67c530da523ed29c0c8ed65b6eeef175
> Local attestation succeeded!
> trusted: file:pipe
> trusted: file:/home/ya0guang/Code/graphene/Runtime/libsysdb.so
> trusted: file:/home/ya0guang/Code/graphene/Runtime/ld-linux-x86-64.so.2
> trusted: file:/home/ya0guang/Code/graphene/Runtime/libc.so.6
> trusted: file:/lib/x86_64-linux-gnu/libc.so.6
> Please input data:this is child  2
> Please input data:this is parent  1
> receive 11 data：I'm a pipe!

It seems debug output still exists. @StanPlatinum and I guess it's because `make clean` doesn't delete some intermediate results.

### Manifest template

```manifest
# your_app manifest file example
#
# This manifest was prepared and tested on Ubuntu 16.04.

loader.argv0_override = "$(APP_NAME)"

# LibOS layer library of Graphene. There is currently only one implementation,
# so it is always set to libsysdb.so. 
loader.preload = "file:$(GRAPHENEDIR)/Runtime/libsysdb.so"

# Show/hide debug log of Graphene ('inline' or 'none' respectively).
loader.debug_type = "$(GRAPHENEDEBUG)"

# Read application arguments directly from the command line. Don't use this on production!
loader.insecure__use_cmdline_argv = 1

# Specify paths to search for libraries. The usual LD_LIBRARY_PATH syntax
# applies. Paths must be in-Graphene visible paths, not host-OS paths (i.e.,
# paths must be taken from fs.mount.xxx.path, not fs.mount.xxx.uri).
loader.env.LD_LIBRARY_PATH = "/lib:$(ARCH_LIBDIR):/usr/$(ARCH_LIBDIR):/usr/lib/$(ARCH_LIBDIR)"

# Mount host-OS directory to required libraries (in 'uri') into in-Graphene
# visible directory /lib (in 'path').
fs.mount.lib.type = "chroot"
fs.mount.lib.path = "/lib"
fs.mount.lib.uri = "file:$(GRAPHENEDIR)/Runtime"

fs.mount.lib2.type = "chroot"
fs.mount.lib2.path = "$(ARCH_LIBDIR)"
fs.mount.lib2.uri = "file:$(ARCH_LIBDIR)"

fs.mount.lib3.type = "chroot"
fs.mount.lib3.path = "/usr$(ARCH_LIBDIR)"
fs.mount.lib3.uri = "file:/usr/$(ARCH_LIBDIR)"

# fs.mount.etc.type = "chroot"
# fs.mount.etc.path = "/etc"
# fs.mount.etc.uri = "file:/etc"

fs.mount."$(APP_NAME)".type = "chroot"
fs.mount."$(APP_NAME)".path = "$(APP_DIR)"
fs.mount."$(APP_NAME)".uri = "file:$(APP_DIR)"

# Set enclave size (somewhat arbitrarily) to 256MB. Recall that SGX v1 requires
# to specify enclave size at enclave creation time.
sgx.enclave_size = "256M"

# Set maximum number of in-enclave threads (somewhat arbitrarily) to 4. Recall
# that SGX v1 requires to specify the maximum number of simultaneous threads at
# enclave creation time.
sgx.thread_num = 6

# Specify all libraries used by $(APP_NAME) and its dependencies (including all libs
# which can be loaded at runtime via dlopen).
sgx.trusted_files.ld = "file:$(GRAPHENEDIR)/Runtime/ld-linux-x86-64.so.2"
sgx.trusted_files.libc = "file:$(GRAPHENEDIR)/Runtime/libc.so.6"
# sgx.trusted_files.libm = "file:$(GRAPHENEDIR)/Runtime/libm.so.6"
# sgx.trusted_files.libdl = "file:$(GRAPHENEDIR)/Runtime/libdl.so.2"
# sgx.trusted_files.librt = "file:$(GRAPHENEDIR)/Runtime/librt.so.1"
# sgx.trusted_files.libutil = "file:$(GRAPHENEDIR)/Runtime/libutil.so.1"
# sgx.trusted_files.libpthread = "file:$(GRAPHENEDIR)/Runtime/libpthread.so.0"
# sgx.trusted_files.libgcc_s = "file:/usr/lib/x86_64-linux-gnu/libgcc_s.so.1"
# sgx.trusted_files.libnss_files = "file:$(ARCH_LIBDIR)/libnss_files.so.2"
# sgx.trusted_files.libnss_mdns4_minimal = "file:$(ARCH_LIBDIR)/libnss_mdns4_minimal.so.2"
$(APP_TRUSTED_LIBS)

# Name Service Switch (NSS) files. Glibc reads these files as part of name-
# service information gathering. For more info, see 'man nsswitch.conf'.
# sgx.allowed_files.nsswitch = "file:/etc/nsswitch.conf"
# sgx.allowed_files.hostconf = "file:/etc/host.conf"
# sgx.allowed_files.resolvconf = "file:/etc/resolv.conf"
# sgx.allowed_files.ethers = "file:/etc/ethers"
# sgx.allowed_files.hosts = "file:/etc/hosts"
# sgx.allowed_files.group = "file:/etc/group"
# sgx.allowed_files.passwd = "file:/etc/passwd"
# sgx.allowed_files.gaiconf = "file:/etc/gai.conf"
```

### Makefile

```Makefile
# Use one of these commands to build the manifest for your app:
#
# - make
# - make DEBUG=1
# - make SGX=1
# - make SGX=1 DEBUG=1
#
# Use `make clean` to remove Graphene-generated files.

THIS_DIR := $(dir $(lastword $(MAKEFILE_LIST)))
APP_DIR ?= .
APP_NAME = pipe

# Relative path to Graphene root and key for enclave signing
GRAPHENEDIR ?= /home/ya0guang/graphene
SGX_SIGNER_KEY ?= $(GRAPHENEDIR)/Pal/src/host/Linux-SGX/signer/enclave-key.pem

ifeq ($(DEBUG),1)
GRAPHENEDEBUG = inline
else
GRAPHENEDEBUG = none
endif

.PHONY: all
all: $(APP_NAME).manifest $(APP_NAME) pal_loader
ifeq ($(SGX),1)
all: $(APP_NAME).token
endif

include $(GRAPHENEDIR)/Scripts/Makefile.configs

# $(APP_NAME) dependencies (generated from ldd). For SGX, the manifest needs to list all the libraries
# loaded during execution, so that the signer can include the file hashes.

# We need to replace Glibc dependencies with Graphene-specific Glibc. The Glibc binaries are
# already listed in the manifest template, so we can skip them from the ldd results.
# GLIBC_DEPS = linux-vdso.so.1 /lib64/ld-linux-x86-64.so.2 libc.so.6 libpthread.so.0
GLIBC_DEPS = linux-vdso.so.1 

# List all the $(APP_NAME) dependencies, besides Glibc libraries
.INTERMEDIATE: $(APP_NAME)-deps
$(APP_NAME)-deps: $(APP_NAME)
	@echo $(patsubst %,-e %,$(GLIBC_DEPS))
	# @ldd $(APP_DIR)/$(APP_NAME) | \
	# 	awk '{if ($$2 =="=>") {print $$1}}' | \
	# 	sort | uniq | grep -v -x $(patsubst %,-e %,$(GLIBC_DEPS)) > $@
	@ldd $(APP_DIR)/$(APP_NAME) | \
		awk '{if ($$2 =="=>") {print $$1}}' | \
		sort | uniq | grep -v -x $(patsubst %,-e %,$(GLIBC_DEPS)) > $@

# Generate manifest rules for $(APP_NAME) dependencies
.INTERMEDIATE: $(APP_NAME)-trusted-libs
$(APP_NAME)-trusted-libs: $(APP_NAME)-deps
	@for F in `cat $(APP_NAME)-deps`; do \
		P=`ldd $(APP_DIR)/$(APP_NAME) | grep $$F | awk '{print $$3; exit}'`; \
		N=`echo $$F | tr --delete '.' | tr --delete '-'`; \
		echo -n "sgx.trusted_files.$$N = \\\"file:$$P\\\"\\\\n"; \
	done > $@

$(APP_NAME).manifest: $(APP_NAME).manifest.template $(APP_NAME)-trusted-libs
	@sed -e 's|$$(GRAPHENEDIR)|'"$(GRAPHENEDIR)"'|g' \
		-e 's|$$(GRAPHENEDEBUG)|'"$(GRAPHENEDEBUG)"'|g' \
		-e 's|$$(APP_DIR)|'"$(APP_DIR)"'|g' \
		-e 's|$$(APP_TRUSTED_LIBS)|'"`cat $(APP_NAME)-trusted-libs`"'|g' \
		-e 's|$$(ARCH_LIBDIR)|'"$(ARCH_LIBDIR)"'|g' \
		$< > $@

# Generate SGX-specific manifest, enclave signature, and token for enclave initialization
$(APP_NAME).manifest.sgx: $(APP_NAME) $(APP_NAME).manifest
	$(GRAPHENEDIR)/Pal/src/host/Linux-SGX/signer/pal-sgx-sign \
		-exec $(APP_NAME) \
		-libpal $(GRAPHENEDIR)/Runtime/libpal-Linux-SGX.so \
		-key $(SGX_SIGNER_KEY) \
		-manifest $(APP_NAME).manifest -output $@

$(APP_NAME).sig: $(APP_NAME).manifest.sgx

$(APP_NAME).token: $(APP_NAME).sig
	$(GRAPHENEDIR)/Pal/src/host/Linux-SGX/signer/pal-sgx-get-token \
		-output $@ -sig $^

$(APP_NAME): $(APP_NAME).c
	@gcc $(APP_NAME).c -o $(APP_DIR)/$@

pal_loader:
	ln -s $(GRAPHENEDIR)/Runtime/pal_loader $@

.PHONY: run
run: all
	./pal_loader ./$(APP_NAME)

.PHONY: clean
clean:
	$(RM) *.manifest *.manifest.sgx *.token *.sig $(APP_NAME) pal_loader OUTPUT

.PHONY: distclean
distclean: clean
```

## Steps to reproduce
<!-- NOTE: please specify the exact commit ID on which you reproduced the issue

### PLEASE ENSURE THAT THE ISSUE REPRODUCES ON THE CURRENT MASTER BRANCH ###

-->

I'm on the latest master branch and graphene is built on Ubuntu 20.04 with kernel 5.4.83-custom installed (for FSGSBASE patching).

## Expected results

No error output.

> Please input data:this is child  2
> Please input data:this is parent  1
> receive 11 data：I'm a pipe!


<!-- ## Additional information -->
<!-- if applicable, uncomment and fill this section -->

--- a/LibOS/shim/src/fs/shim_fs.c
+++ b/LibOS/shim/src/fs/shim_fs.c
@@ -97,14 +97,14 @@ static int __mount_root(struct shim_dentry** root) {
 
     ret = toml_string_in(g_manifest_root, "fs.root.type", &fs_root_type);
     if (ret < 0) {
-        debug("Cannot parse \'fs.root.type\' (the value must be put in double quotes!)\n");
+        debug("Cannot parse 'fs.root.type' (the value must be put in double quotes!)\n");
         ret = -EINVAL;
         goto out;
     }
 
     ret = toml_string_in(g_manifest_root, "fs.root.uri", &fs_root_uri);
     if (ret < 0) {
-        debug("Cannot parse \'fs.root.uri\' (the value must be put in double quotes!)\n");
+        debug("Cannot parse 'fs.root.uri' (the value must be put in double quotes!)\n");
         ret = -EINVAL;
         goto out;
     }
@@ -163,19 +163,19 @@ static int __mount_one_other(toml_table_t* mount) {
 
     toml_raw_t mount_type_raw = toml_raw_in(mount, "type");
     if (!mount_type_raw) {
-        debug("Cannot find \'fs.mount.%s.type\'\n", key);
+        debug("Cannot find 'fs.mount.%s.type'\n", key);
         return -EINVAL;
     }
 
     toml_raw_t mount_path_raw = toml_raw_in(mount, "path");
     if (!mount_path_raw) {
-        debug("Cannot find \'fs.mount.%s.path\'\n", key);
+        debug("Cannot find 'fs.mount.%s.path'\n", key);
         return -EINVAL;
     }
 
     toml_raw_t mount_uri_raw = toml_raw_in(mount, "uri");
     if (!mount_uri_raw) {
-        debug("Cannot find \'fs.mount.%s.uri\'\n", key);
+        debug("Cannot find 'fs.mount.%s.uri'\n", key);
         return -EINVAL;
     }
 
@@ -185,21 +185,21 @@ static int __mount_one_other(toml_table_t* mount) {
 
     ret = toml_rtos(mount_type_raw, &mount_type);
     if (ret < 0) {
-        debug("Cannot parse \'fs.mount.%s.type\' (the value must be put in double quotes!)\n", key);
+        debug("Cannot parse 'fs.mount.%s.type' (the value must be put in double quotes!)\n", key);
         ret = -EINVAL;
         goto out;
     }
 
     ret = toml_rtos(mount_path_raw, &mount_path);
     if (ret < 0) {
-        debug("Cannot parse \'fs.mount.%s.path\' (the value must be put in double quotes!)\n", key);
+        debug("Cannot parse 'fs.mount.%s.path' (the value must be put in double quotes!)\n", key);
         ret = -EINVAL;
         goto out;
     }
 
     ret = toml_rtos(mount_uri_raw, &mount_uri);
     if (ret < 0) {
-        debug("Cannot parse \'fs.mount.%s.uri\' (the value must be put in double quotes!)\n", key);
+        debug("Cannot parse 'fs.mount.%s.uri' (the value must be put in double quotes!)\n", key);
         ret = -EINVAL;
         goto out;
     }
@@ -209,13 +209,13 @@ static int __mount_one_other(toml_table_t* mount) {
     if (!strcmp(mount_path, "/")) {
         debug("Root mount / already exists, verify that there are no duplicate mounts in manifest\n"
               "(note that root / is automatically mounted in Graphene and can be changed via "
-              "\'fs.root\' manifest entry).\n");
+              "'fs.root' manifest entry).\n");
         ret = -EEXIST;
         goto out;
     }
 
     if (!strcmp(mount_path, ".") || !strcmp(mount_path, "..")) {
-        debug("Mount points \'.\' and \'..\' are not allowed, remove them from manifest.\n");
+        debug("Mount points '.' and '..' are not allowed, remove them from manifest.\n");
         ret = -EINVAL;
         goto out;
     }
@@ -292,7 +292,7 @@ int init_mount(void) {
     char* fs_start_dir = NULL;
     ret = toml_string_in(g_manifest_root, "fs.start_dir", &fs_start_dir);
     if (ret < 0) {
-        debug("Can't parse \"fs.start_dir\" (note that the value must be put in double quotes)!\n");
+        debug("Can't parse 'fs.start_dir' (note that the value must be put in double quotes)!\n");
         return ret;
     }
 
@@ -301,7 +301,7 @@ int init_mount(void) {
         ret = path_lookupat(NULL, fs_start_dir, 0, &dent, NULL);
         free(fs_start_dir);
         if (ret < 0) {
-            debug("Invalid \"fs.start_dir\" in manifest.\n");
+            debug("Invalid 'fs.start_dir' in manifest.\n");
             return ret;
         }
         lock(&g_process.fs_lock);

--- a/LibOS/shim/src/fs/shim_fs.c
+++ b/LibOS/shim/src/fs/shim_fs.c
@@ -214,6 +214,12 @@ static int __mount_one_other(toml_table_t* mount) {
         goto out;
     }
 
+    if (!strcmp(mount_path, ".") || !strcmp(mount_path, "..")) {
+        debug("Mount points \'.\' and \'..\' are not allowed, remove them from manifest.\n");
+        ret = -EINVAL;
+        goto out;
+    }
+
     if ((ret = mount_fs(mount_type, mount_uri, mount_path, NULL, NULL, 1)) < 0) {
         debug("Mounting %s on %s (type=%s) failed (%d)\n", mount_uri, mount_path, mount_type,
               -ret);

